apps/extension/src/storage/onboard.ts

```diff
@@ -0,0 +1,51 @@
+import { ExtensionStorage } from '@penumbra-zone/storage/chrome/base';
+import { localExtStorage } from '@penumbra-zone/storage/chrome/local';
+import { SessionStorageState } from '@penumbra-zone/storage/chrome/session';
+import { LocalStorageState } from '@penumbra-zone/storage/chrome/types';
+
+export const waitForStorageKey = async <
+  S extends LocalStorageState | SessionStorageState,
+  K extends keyof S,
+  I extends keyof S[K],
+>(
+  storage: ExtensionStorage<S>,
+  storageKey: K & string,
+  innerKey?: I,
+): Promise<I extends undefined ? S[K] : S[K][I]> => {
+  type Stored = I extends undefined ? S[K] : S[K][I];
+  const { promise, resolve, reject } = Promise.withResolvers<Stored>();
+
+  const storedObject: S[K] = await storage.get(storageKey);
+  const storedValue = (innerKey != null ? storedObject[innerKey] : storedObject) as
+    | Stored
+    | undefined;
+  if (storedValue != null) resolve(storedValue);
+  else {
+    const storageListener = (changes: Record<string, { newValue?: unknown }>) => {
```

Comment:
Instead of manually typing the Record type for changes, think we should have a type we can import here from `packages/storage/src/chrome/base.ts`

Reply:
i spent some time on this, but it gets complex quickly, so i'm no longer implementing this reuseable method. considering it out of scope.